tensorflow/core/util/gpu_device_functions.h

```diff
@@ -489,31 +529,32 @@ __device__ T CudaAtomicCasHelper(T* ptr, F accumulate) {
   } while (assumed != old);
   return old;
 }
+CREATE_CUDA_DEVICE_FUNCTION_ALIAS(GpuAtomicCasHelper, CudaAtomicCasHelper);
```

Comment:
Is it valid to add the alias before the template specializations?

Reply:
I used to have this concern as well. Looking at what the macro does, it does a text replacement to a perfect forwarding signature of the new function name. So the question ended up as whether the perfect forwarding declaration is location relevant. 

I'd say it is not because when compiler does the actual forwarding, it needs to match up the types and number of arguments. The process of matching up the correct function signature should be as the same as if the compiler resolves any other template functions.

Reply:
I think you are correct, the compiler has seen all template specializations when it instantiates the aliasing template.